hoomd/md/external/wall.py

```diff
@@ -0,0 +1,577 @@
+# Copyright (c) 2009-2021 The Regents of the University of Michigan
+# This file is part of the HOOMD-blue project, released under the BSD 3-Clause
+# License.
+
+r"""Wall potentials.
+
+Wall potentials add forces to any particles within a certain distance,
+:math:`r_{\mathrm{cut}}`, of each wall surface. In the extrapolated mode, all
+particles outside of the wall boundary are included as well.
+"""
+
+from hoomd.md import force
+from hoomd.data.array_view import _ArrayViewWrapper
+from hoomd.md import _md
+import hoomd
+
+
+def _to_md_cpp_wall(wall):
+    if isinstance(wall, hoomd.wall.Sphere):
+        return _md.SphereWall(wall.radius, wall.origin.to_base(), wall.inside,
+                              wall.open)
+    if isinstance(wall, hoomd.wall.Cylinder):
+        return _md.CylinderWall(wall.radius, wall.origin.to_base(),
+                                wall.axis.to_base(), wall.inside, wall.open)
+    if isinstance(wall, hoomd.wall.Plane):
+        return _md.PlaneWall(wall.origin.to_base(), wall.normal.to_base(),
+                             wall.open)
+    raise TypeError(f"Unknown wall type encountered {type(wall)}.")
+
+
+class _WallArrayViewFactory:
+
+    def __init__(self, cpp_wall_potential, wall_type):
+        self.cpp_obj = cpp_wall_potential
+        self.func_name = {
+            hoomd.wall.Sphere: "get_sphere_list",
+            hoomd.wall.Cylinder: "get_cylinder_list",
+            hoomd.wall.Plane: "get_plane_list"
+        }[wall_type]
+
+    def __call__(self):
+        return getattr(self.cpp_obj.field, self.func_name)()
+
+
+class WallPotential(force.Force):
+    r"""Generic wall potential.
+
+    Warning:
+        `WallPotential` should not be used directly.  It is a base class that
+        provides features and documentation common to all standard wall
+        potentials.
+
+    All wall potential classes specify that a given potential energy and
+    force be computed on all particles in the system when the signed cutoff
+    distance is near the wall surface: :math:`r < r_{\mathrm{cut}}`.
+    The force :math:`\vec{F}` is in the direction of :math:`\vec{r}`, the vector
```

Comment:
I think the directions are reversed here. This says that a particle that is very close to a LJ wall (e.g., in the repulsive region) would have a very large force acting on it that points toward the wall.

Reply:
Forces have signs, so stating "in the direction" here is ambiguous. I will push a change to correct this wording to be unambiguous. We also need to reference the vector from the wall to the point as `\vec{d}` because `\vec{r}` is used for particle positions.

Reply:
Let me know what you think of the reworded overview.

Reply:
Much better.